### ISSUE TITLE
[flutter_adaptive_scaffold] [go_router] Please create documentation for using the AdaptiveScaffold and go_router together

### ISSUE THREAD

**Issue by hanskokx** [flutter/flutter#129850]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I would like an application with nested navigation that is responsive.  See the following example:

![image](https://github.com/flutter/flutter/assets/1911919/38a11b8a-6b79-42da-b613-edab1353ea22)

This has proven to be exceptionally challenging, given that there is no documentation around how to use the `secondaryBody`, especially when given a sub-route.

### Proposal

After hacking away at the problem for a while, I came up with [this](https://github.com/hanskokx/flutter_adaptive_scaffold_example) solution.  I believe there's _much_ room for improvement, however. I would like to see some documentation surrounding this use case, as it seems like it would be a very common way of using this widget.

**Comment by hanskokx** [comment]
I've been thinking about my implementation for a while now. There are a couple things that bother me:

- Tapping on a navigation item then tapping something on the list shows the details for that item. However, navigating to a different navigation item and back loses the state of the initial route - so you'd have to navigate back to the details.
- This solution uses both the `AdaptiveScaffold` and `AdaptiveLayout`. In my mind, this is excessive. I should be able to use just the `AdaptiveScaffold` _or_ the `AdaptiveLayout` while still maintaining state. (Much like the Cupertino widgets allow).

Let me know if I'm talking crazy talk, or if you need more information.

**Comment by Hannnes1** [comment]
> I've been thinking about my implementation for a while now. There are a couple things that bother me:
> 
>     * Tapping on a navigation item then tapping something on the list shows the details for that item. However, navigating to a different navigation item and back loses the state of the initial route - so you'd have to navigate back to the details.
> 
>     * This solution uses both the `AdaptiveScaffold` and `AdaptiveLayout`. In my mind, this is excessive. I should be able to use just the `AdaptiveScaffold` _or_ the `AdaptiveLayout` while still maintaining state. (Much like the Cupertino widgets allow).
> 
> 
> Let me know if I'm talking crazy talk, or if you need more information.

I just found this issue while looking for an example myself. It seems like your first point is fixed, right?

**Comment by fredgrott** [comment]
It's about to get worse, does anyone realize the MD3 Canonical Layouts are anti-layouts in that its the combo of Flutter Adaptive Scaffold, go router, smart Panes, and smart listDetail and smart card grids that actually form the layout but it will not have a predefined layout class??

**Comment by Hannnes1** [comment]
> It's about to get worse, does anyone realize the MD3 Canonical Layouts are anti-layouts in that its the combo of Flutter Adaptive Scaffold, go router, smart Panes, and smart listDetail and smart card grids that actually form the layout but it will not have a predefined layout class??

Yeah, it's quite a complicated problem. We have abandoned this design completely for now, since we couldn't figure out a good way to implement it.
But I realize that this package is still in early development, so I expect things like this to become clearer along the way.

**Comment by fredgrott** [comment]
I came up with a possible solution it involves passing the Go Route(Screen) to the every body breakpoint in the adaptive scaffold and then that screen will use layouts:
   small mobile Pane that morphs to two pane
  small mobile Pane that morphs to split two pane

the screen layouts will do the hasHinge unfolding and responsive stuff via LayoutBuilder.

If that does indeed work then it would seem that all the secondBody slots will be chalked up 
to implement error as they would then not be needed.

And of course I should make the attempt to implement as one CanonicalLayout that 
has a Boolean isTwoPane when false would then be SplitPane.

@Hannnes1 this might meet your requirements as well, what do think is the approach do-able in your view?

**Comment by Hannnes1** [comment]
> I came up with a possible solution it involves passing the Go Route(Screen) to the every body breakpoint in the adaptive scaffold and then that screen will use layouts:
>    small mobile Pane that morphs to two pane
>   small mobile Pane that morphs to split two pane
> 
> the screen layouts will do the hasHinge unfolding and responsive stuff via LayoutBuilder.
> 
> If that does indeed work then it would seem that all the secondBody slots will be chalked up 
> to implement error as they would then not be needed.
> 
> And of course I should make the attempt to implement as one CanonicalLayout that 
> has a Boolean isTwoPane when false would then be SplitPane.
> 
> @Hannnes1 this might meet your requirements as well, what do think is the approach do-able in your view?

I haven't had time to test it yet, but that might work

**Comment by fredgrott** [comment]
For those that cannot wait on Flutter SDK team, steps to do it:

1. To make second body slots work, implement
          -windowsize enum
         -body slot enum, note secondbody slots should  only exist for expanded and higher
        -buildcontext ext to determine correct window size per width
       -body slot inherited widget to pass the body slot enum

    *body slot childs get wrapped by inherited widget so that the screen layout is a smart canonical layout that adopts to delivering the right layout to body slot and because its the same screen for all body slots maintains the correct state...even in desktop when changing window sizes and yet delivers the right transition to foldables.
2. Rewrite the adaptive scaffold using the slot and adaptive layout APIs and supply the correct GoRouter object for StatefulShell Route navigation index...

3. Last step rewrite animated indexed stack per these directions:
https://github.com/flutter/flutter/issues/48217#issuecomment-1636932641

That corrects several problems in the framework to make page transition animation work with indexed  stacks.


4.  Now you are ready to write your canonical layouts

Or if you do want to do that find my substack....and yes its under my nickname...

Its called lets party as we know how to support iPhone foldables before they show up in the marketplace!

**Comment by martijn00** [comment]
@fredgrott I'm working on a series of PR's to update to the latest M3 specs and to more align with the new Compose libs for this. Could you provide feedback on those as you are deep into your research on this?

**Comment by martijn00** [comment]
@fredgrott  Besides the GoRouter PR i've also opened this one: https://github.com/flutter/packages/pull/7428

It has the base values need to make adjustments to UI which would come in another PR.

**Comment by snapsl** [comment]
The current hinge position calculation in AdaptiveLayout does not work when it is nested inside an AdaptiveScaffold.
